components/supervisor/pkg/terminal/service.go

```diff
@@ -168,18 +235,58 @@ func (srv *MuxTerminalService) Listen(req *api.ListenTerminalRequest, resp api.T
 		buf := make([]byte, 4096)
 		for {
 			n, err := stdout.Read(buf)
+			if err == io.EOF {
+				break
+			}
 			if err != nil {
 				errchan <- err
 				return
 			}
 
-			// TODO(cw): find out how to separate stdout/stderr
-			err = resp.Send(&api.ListenTerminalResponse{Output: &api.ListenTerminalResponse_Stdout{Stdout: buf[:n]}})
+			err = resp.Send(&api.ListenTerminalResponse{Output: &api.ListenTerminalResponse_Data{Data: buf[:n]}})
 			if err != nil {
 				errchan <- err
 				return
 			}
 		}
+
+		state, err := term.Wait()
+		if err != nil {
+			errchan <- err
+			return
+		}
+
+		err = resp.Send(&api.ListenTerminalResponse{Output: &api.ListenTerminalResponse_ExitCode{ExitCode: int32(state.ExitCode())}})
+		if err != nil {
+			errchan <- err
+			return
+		}
+		errchan <- io.EOF
+	}()
+	go func() {
```

Comment:
it is probably better to pool it once for a terminal instead of for each listener

Reply:
currently there is only one client (IDE backend) which then distributes event between all windows